components/builder-api/src/server/resources/origins.rs

```diff
@@ -1158,14 +1202,121 @@ fn list_origin_invitations(req: HttpRequest,
+        }
+        Err(err) => {
+            debug!("{}", err);
+            Error::DieselError(err).into()
+        }
+    }
+}
+
+#[allow(clippy::needless_pass_by_value)]
+fn update_origin_member_role(req: HttpRequest,
+                             path: Path<(String, String)>,
+                             req_role: Query<Role>,
+                             state: Data<AppState>)
+                             -> HttpResponse {
+    let (origin, username) = path.into_inner();
+    let target_role = match OriginMemberRole::from_str(&req_role.role) {
+        Ok(r) => {
+            debug!("role {}", r);
+            r
+        }
+        Err(err) => {
+            debug!("{}", err);
+            return HttpResponse::new(StatusCode::UNPROCESSABLE_ENTITY);
+        }
+    };
+
+    // Account id of the user making the request
+    let account_id = match authorize_session(&req,
+                                             Some(&origin),
+                                             RoleAction::Write,
+                                             Some(OriginMemberRole::Maintainer))
+    {
+        Ok(session) => session.get_id(),
+        Err(err) => return err.into(),
+    };
+
+    let conn = match state.db.get_conn().map_err(Error::DbError) {
+        Ok(conn_ref) => conn_ref,
+        Err(err) => return err.into(),
+    };
+
+    // We cannot allow a user to escalate to Owner. That must be done via Origin owner transfer.
+    if target_role == OriginMemberRole::Owner {
+        return HttpResponse::new(StatusCode::FORBIDDEN);
+    }
+
+    // The account id of the user being requested
+    let (target_user_id, _) = match Account::get(&username, &*conn) {
+        Ok(account) => (account.id, account.name),
+        Err(err) => {
+            debug!("{}", err);
+            return Error::DieselError(err).into();
+        }
+    };
+
+    // We cannot allow a user to change their own role
+    if account_id as i64 == target_user_id {
+        return HttpResponse::new(StatusCode::FORBIDDEN);
+    }
+
+    // We cannot allow a user to change the role of the origin owner
+    if check_origin_owner(&req, target_user_id as u64, &origin).unwrap_or(false) {
+        return HttpResponse::new(StatusCode::FORBIDDEN);
+    }
+
+    match OriginMember::update_member_role(&origin, target_user_id as i64, &*conn, target_role) {
+        Ok(_) => HttpResponse::new(StatusCode::NO_CONTENT),
+        Err(err) => {
+            debug!("{}", err);
+            Error::DieselError(err).into()
+        }
+    }
+}
+
 #[allow(clippy::needless_pass_by_value)]
 fn transfer_origin_ownership(req: HttpRequest,
                              path: Path<(String, String)>,
                              state: Data<AppState>)
                              -> HttpResponse {
     let (origin, user) = path.into_inner();
 
-    let session = match authorize_session(&req, None) {
+    let session = match authorize_session(&req, None, RoleAction::Write, None) {
```

Comment:
Should this require the owner role?

Reply:
This does require owner role - all of the previous deep guards for authorization have remained in place.  This is an early exit which should help with the need to check if the requester is the owner which happens on 1324